File: qlat-utils/pylib/qlat_utils/qar.py
import qlat_utils.c as c

from qlat_utils.c import get_qar_multi_vol_max_size
from qlat_utils.c import does_regular_file_exist_qar
from qlat_utils.c import does_file_exist_qar
from qlat_utils.c import qcat
from qlat_utils.c import qcat_bytes
from qlat_utils.c import qar_create
from qlat_utils.c import qar_extract
from qlat_utils.c import qcopy_file
from qlat_utils.c import list_qar

from qlat_utils.timer import *

def does_regular_file_exist_qar_sync_node(path):
    if get_num_node() != 1:
        import cqlat as c
        return c.does_regular_file_exist_qar_sync_node(path)
    return does_regular_file_exist_qar(path)

def does_file_exist_qar_sync_node(path):
    if get_num_node() != 1:
        import cqlat as c
        return c.does_file_exist_qar_sync_node(path)
    return does_file_exist_qar(path)

def qcat_sync_node(path):
    if get_num_node() != 1:
        import cqlat as c
        return c.qcat_sync_node(path)
    return qcat(path)

def qcat_bytes_sync_node(path):
    if get_num_node() != 1:
        import cqlat as c
        return c.qcat_bytes_sync_node(path)
    return qcat_bytes(path)

@timer
def qar_create_info(path_qar, path_folder, *, is_remove_folder_after = False):
    if get_num_node() != 1:
        import cqlat as c
        return c.qar_create_info(path_qar, path_folder, is_remove_folder_after)
    return qar_create(path_qar, path_folder, is_remove_folder_after = is_remove_folder_after)

@timer
def qar_extract_info(path_qar, path_folder, *, is_remove_qar_after = False):
    if get_num_node() != 1:
        import cqlat as c
        return c.qar_extract_info(path_qar, path_folder, is_remove_qar_after)
    return qar_extract(path_qar, path_folder, is_remove_qar_after = is_remove_qar_after)

@timer
def qcopy_file_info(path_src, path_dst):
    if get_num_node() != 1:
        import cqlat as c
        return c.qcopy_file_info(path_src, path_dst)
    return qcopy_file(path_src, path_dst)

File: qlat-utils/pylib/cqlat_utils/qar.cpp
#include "lib.h"

#include <qlat-utils/qar-cache.h>

EXPORT(does_file_exist_qar, {
  using namespace qlat;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_path)) {
    return NULL;
  }
  const std::string path = py_convert_data<std::string>(p_path);
  const bool ret = does_file_exist_qar(path);
  return py_convert(ret);
})

EXPORT(does_file_or_directory_exist_qar, {
  using namespace qlat;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_path)) {
    return NULL;
  }
  const std::string path = py_convert_data<std::string>(p_path);
  const bool ret = does_file_or_directory_exist_qar(path);
  return py_convert(ret);
})

EXPORT(get_qar_multi_vol_max_size, {
  using namespace qlat;
  PyObject* p_size = NULL;
  if (!PyArg_ParseTuple(args, "|O", &p_size)) {
    return NULL;
  }
  if (NULL != p_size) {
    const long size = py_convert_data<long>(p_size);
    get_qar_multi_vol_max_size() = size;
  }
  return py_convert(get_qar_multi_vol_max_size());
})

EXPORT(qar_create, {
  using namespace qlat;
  PyObject* p_path_qar = NULL;
  PyObject* p_path_folder = NULL;
  bool is_remove_folder_after = false;
  if (!PyArg_ParseTuple(args, "OO|b", &p_path_qar, &p_path_folder, &is_remove_folder_after)) {
    return NULL;
  }
  const std::string path_qar = py_convert_data<std::string>(p_path_qar);
  const std::string path_folder = py_convert_data<std::string>(p_path_folder);
  const int ret = qar_create(path_qar, path_folder, is_remove_folder_after);
  return py_convert(ret);
})

EXPORT(qar_extract, {
  using namespace qlat;
  PyObject* p_path_qar = NULL;
  PyObject* p_path_folder = NULL;
  bool is_remove_qar_after = false;
  if (!PyArg_ParseTuple(args, "OO|b", &p_path_qar, &p_path_folder,
                        &is_remove_qar_after)) {
    return NULL;
  }
  const std::string path_qar = py_convert_data<std::string>(p_path_qar);
  const std::string path_folder = py_convert_data<std::string>(p_path_folder);
  const int ret = qar_extract(path_qar, path_folder, is_remove_qar_after);
  return py_convert(ret);
})

EXPORT(qcopy_file, {
  using namespace qlat;
  PyObject* p_path_src = NULL;
  PyObject* p_path_dst = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_path_src, &p_path_dst)) {
    return NULL;
  }
  const std::string path_src = py_convert_data<std::string>(p_path_src);
  const std::string path_dst = py_convert_data<std::string>(p_path_dst);
  const int ret = qcopy_file(path_src, path_dst);
  return py_convert(ret);
})

EXPORT(list_qar, {
  using namespace qlat;
  PyObject* p_path_qar = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_path_qar)) {
    return NULL;
  }
  const std::string path_qar = py_convert_data<std::string>(p_path_qar);
  return py_convert(list_qar(path_qar));
})

EXPORT(qcat, {
  using namespace qlat;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_path)) {
    return NULL;
  }
  const std::string path = py_convert_data<std::string>(p_path);
  const std::string ret = qcat(path);
  return py_convert(ret);
})

EXPORT(qcat_bytes, {
  using namespace qlat;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_path)) {
    return NULL;
  }
  const std::string path = py_convert_data<std::string>(p_path);
  const std::string ret = qcat(path);
  return py_convert(get_data(ret));
})
